cli/src/diff_util.rs

```diff
@@ -1128,27 +1153,73 @@ pub fn show_diff_summary(
     mut tree_diff: TreeDiffStream,
     path_converter: &RepoPathUiConverter,
 ) -> io::Result<()> {
+    #[derive(Debug)]
+    enum Type {
+        Copy(String, String),
+        Rename(String, String),
+        Modification(String),
+        Addition(String),
+        Deletion(String),
+    }
+
+    let mut file_modifications = Vec::<Type>::new();
+    let mut unresolved_renames = HashMap::<String, usize>::new();
     async {
         while let Some(TreeDiffEntry {
-            source: _,
-            target: repo_path,
+            source: before_path,
+            target: after_path,
             value: diff,
         }) = tree_diff.next().await
         {
             let (before, after) = diff.unwrap();
-            let ui_path = path_converter.format_file_path(&repo_path);
-            if before.is_present() && after.is_present() {
-                writeln!(formatter.labeled("modified"), "M {ui_path}")?;
-            } else if before.is_absent() {
-                writeln!(formatter.labeled("added"), "A {ui_path}")?;
+            let before_ui_path = path_converter.format_file_path(&before_path);
+            let after_ui_path = path_converter.format_file_path(&after_path);
+            if before_path != after_path {
+                if let Some(&idx) = unresolved_renames.get(&before_ui_path) {
+                    if let Some(Type::Deletion(_)) = file_modifications.get(idx) {
+                        file_modifications[idx] = Type::Rename(before_ui_path, after_ui_path);
+                    }
+                } else {
+                    unresolved_renames.insert(before_ui_path.clone(), file_modifications.len());
+                    file_modifications.push(Type::Copy(before_ui_path, after_ui_path));
+                }
             } else {
-                // `R` could be interpreted as "renamed"
-                writeln!(formatter.labeled("removed"), "D {ui_path}")?;
+                match (before.is_present(), after.is_present()) {
+                    (true, true) => file_modifications.push(Type::Modification(before_ui_path)),
+                    (false, true) => file_modifications.push(Type::Addition(before_ui_path)),
+                    (true, false) => {
+                        if let Some(&idx) = unresolved_renames.get(&before_ui_path) {
+                            if let Some(Type::Copy(before, after)) = file_modifications.get(idx) {
+                                file_modifications[idx] =
+                                    Type::Rename(before.clone(), after.clone());
+                            }
+                        } else {
+                            unresolved_renames
+                                .insert(before_ui_path.clone(), file_modifications.len());
+                            file_modifications.push(Type::Deletion(before_ui_path));
+                        }
+                    }
+                    _ => {}
```

Comment:
nit: make this explicitly `(false, false)` and make it call `unreachable!()` to clarify that it shouldn't happen (right?)

Reply:
done